designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,652 @@
+
+# Cert-manager configuration
+certmanager:
+  enabled: false
+  issuerName: "letsencrypt-prod"
+  commonName: "example.com"
+  dnsName: "example.com"
+
+# Network policy configuration sourced from the `config/network_policy`
+networkPolicy:
+  enabled: false
+
+# Prometheus configuration
+prometheus:
+  enabled: false
+
+# Sample configuration sourced from the `config/samples`
+samples:
+  install: true
+
+# Manager configuration sourced from the `config/manager`
+manager:
+  replicas: 1
+  image:
+    repository: "controller"
+    tag: "latest"
+  resources:
+    limits:
+      cpu: 100m
+      memory: 128Mi
+    requests:
+      cpu: 100m
+      memory: 64Mi
+
+# Metrics configuration sourced from the `config/metrics`
+metrics:
+  enabled: true
+
+# Leader election configuration sourced from the `config/leader_election`
+leaderElection:
+  enabled: true
+  role: "leader-election-role"
+  rolebinding: "leader-election-rolebinding"
+
+
+# Controller Manager configuration sourced from the `config/manager`
+controllerManager:
+  manager:
+    args:
+    - --metrics-bind-address=:8443
+    - --leader-elect
+    - --health-probe-bind-address=:8081
+    containerSecurityContext:
+      allowPrivilegeEscalation: false
+      capabilities:
+        drop:
+        - ALL
+    image:
+      repository: controller
+      tag: latest
+    resources:
+      limits:
+        cpu: 500m
+        memory: 128Mi
+      requests:
+        cpu: 10m
+        memory: 64Mi
+  replicas: 1
+  serviceAccount:
+    annotations: {}
+
+# Kubernetes cluster domain configuration
+kubernetesClusterDomain: cluster.local
+
+# Metrics service configuration sourced from the `config/metrics`
+metricsService:
+  ports:
+  - name: https
+    port: 8443
+    protocol: TCP
+    targetPort: 8443
+  type: ClusterIP
+
+# Webhook service configuration sourced from the `config/webhook`
+webhookService:
+  ports:
+  - port: 443
+    protocol: TCP
+    targetPort: 9443
+  type: ClusterIP
+```
+
+### Optional configurations should be disabled by default
+
+The HelmChart plugin should not scaffold optional options enabled
+when those are scaffolded as disabled by the default implementation
+of `kustomize/v2` and consequently the `go/v4` plugin used by default. Example:
+
+The dependency on Cert-Manager is disabled by default.
+
+```yaml
+From config/default/kusyomization.yaml
+# [CERTMANAGER] To enable cert-manager, uncomment all sections with 'CERTMANAGER'. 'WEBHOOK' components are required.
+#- ../certmanager
+```
+
+Therefore, by default the `values.yaml` should be scaffolded with:
+
+```
+# Cert-manager configuration
+certmanager:
+  enabled: false
+```
+
+### Namespace Creation
+
+For Golang projects, the namespace will always be created, and the project will be deployed within it.
+For the Helm chart, we should:
+
+```
+# templates/manager/namespace.yaml
+{{- if .Values.namespace.create }}
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{ .Values.namespace.name }}
+{{- end }}
+```
```

Comment:
Remove :

There are a few things that stand out right off the bat, like the notes on namespace substitution. In Helm, a single release can only be in a single namespace. The install and upgradesubcommands work like kubectl regarding namespace: they assume the current namespace, unless -n, --namespace is provided. If a namespace is provided this way, and the namespace doesn't already exist, the install subcommand will fail unless --create-namespace flag is also used. Because of this, neither Helm chart values.yaml or templates contain a namespace key. So that part can be removed from the proposal.